lib/spack/spack/compilers/msvc.py

```diff
@@ -78,19 +150,31 @@ def __init__(self, *args, **kwargs):
         new_pth = [pth if pth else get_valid_fortran_pth(args[0].version) for pth in args[3]]
         args[3][:] = new_pth
         super(Msvc, self).__init__(*args, **kwargs)
-        if os.getenv("ONEAPI_ROOT"):
+        # To use the MSVC compilers, VCVARS must be invoked
+        # VCVARS is located at a fixed location, referencable
+        # idiomatically by the following relative path from the
+        # compiler.
+        # Spack first finds the compilers via VSWHERE
+        # and stores their path, but their respective VCVARS
+        # file must be invoked before useage.
+        env_cmds = []
+        compiler_root = os.path.join(self.cc, "../../../../../../..")
+        self.vcvarsall = os.path.join(compiler_root, "Auxiliary", "Build", "vcvars64.bat")
+        # get current platform architecture and format for vcvars argument
+        arch = spack.platforms.real_host().default.lower()
+        arch = arch.replace("-", "_")
+        self.vcvars_call = VCVarsInvocation(self.vcvarsall, arch, self.msvc_version)
+        env_cmds.append(self.vcvars_call)
+        # Below is a check for a valid fortran path
+        if args[3][2]:
```

Comment:
`args[3]` is the `paths` argument to the base compiler class, and the 3rd `[2]` entry in that list is for the path to the `fc` compiler. I don't see any reason it can't have the same signature as the base compiler class except that it would require the `Msvc` class to be updated each time the base compiler constructor signature is. Let me know if that's an acceptable direction.

Reply:
I think either one has drawbacks in terms of clarity (e.g. detailing all the args just to interfere with one), although this approach should be clarified: starting with line 148, I suggest explicitly pulling out and calling that positional argument "paths", and adding a comment here about how the list of paths is interpreted in the base class.

Reply:
Sounds good to me!